// jwt-authorized/build/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenAuthServiceCallable = exports.loadThroughDi = exports.tokenConfigGenerator = exports.TokenUser = exports.HeaderAuthTokenExtractor = exports.TokenAuthService = void 0;
var loaders_1 = require("./loaders");
Object.defineProperty(exports, "loadThroughDi", { enumerable: true, get: function () { return loaders_1.loadThroughDi; } });
Object.defineProperty(exports, "tokenAuthServiceCallable", { enumerable: true, get: function () { return loaders_1.tokenAuthServiceCallable; } });
var TokenAuthService_1 = __importDefault(require("./services/TokenAuthService"));
exports.TokenAuthService = TokenAuthService_1.default;
var HeaderAuthTokenExtractor_1 = __importDefault(require("./services/HeaderAuthTokenExtractor"));
exports.HeaderAuthTokenExtractor = HeaderAuthTokenExtractor_1.default;
var models_1 = require("./models");
Object.defineProperty(exports, "TokenUser", { enumerable: true, get: function () { return models_1.TokenUser; } });
var tokenConfigGenerator_1 = __importDefault(require("./config/tokenConfigGenerator"));
exports.tokenConfigGenerator = tokenConfigGenerator_1.default;
exports.default = loaders_1.loadThroughDi;
