// saylo/build/src/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Logger = /** @class */ (function () {
    function Logger(onOffMap) {
        this.log = (onOffMap === null || onOffMap === void 0 ? void 0 : onOffMap.log) ? this.turnOn('log') : this.turnOff('log');
        this.debug = (onOffMap === null || onOffMap === void 0 ? void 0 : onOffMap.debug) ? this.turnOn('debug') : this.turnOff('debug');
        this.error = (onOffMap === null || onOffMap === void 0 ? void 0 : onOffMap.error) ? this.turnOn('error') : this.turnOff('error');
    }
    Logger.prototype.turnOn = function (level) {
        var levels = level ? [level] : ['log', 'debug', 'error'];
        return this.turn(levels, console.log.bind(console));
    };
    Logger.prototype.turnOff = function (level) {
        var levels = level ? [level] : ['log', 'debug', 'error'];
        return this.turn(levels, function () { return undefined; });
    };
    Logger.prototype.turn = function (levels, l) {
        var _this = this;
        levels.map(function (level) {
            if (level === 'log') {
                _this.log = l;
            }
            else if (level === 'debug') {
                _this.debug = l;
            }
            else if (level === 'error') {
                _this.error = l;
            }
        });
        return l;
    };
    return Logger;
}());
exports.default = Logger;
